### PR TITLE
Initial plugins and writers for AIE and NOC profiling

### PR DESCRIPTION
**Pull request by pgschuey**

### CODE REVIEW

File: src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+          uint64_t writeMinLatency   = 24;
+          uint64_t writeMaxLatency   = 123;
+          values.push_back(writeByteCount);
+          values.push_back(writeBurstCount);
+          values.push_back(writeTotalLatency);
+          values.push_back(writeMinLatency);
+          values.push_back(writeMaxLatency);
+
+          // Add sample to dynamic database
+	        (db->getDynamicInfo()).addNOCSample(index, timestamp, cellName, values);
```
Comment (keryell) [inline src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp:129]:
```suggestion
            db->getDynamicInfo().addNOCSample(index, timestamp, cellName, values);
```
Also do not mess up with TABs.

File: src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+      xclClose(handle);
+      ++index;
+      handle = xclOpen(index, "/dev/null", XCL_INFO);
+    }
+
+    // Get polling interval (in msec)
+    mPollingInterval = xrt_core::config::get_noc_profile_interval_ms();
+
+    // Start the NOC profiling thread
+    mPollingThread = std::thread(&NOCProfilingPlugin::pollNOCCounters, this);
```
Comment (keryell) [inline src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp:62]: Do not use obfuscated API.
Furthermore, you should probably use `std::async` instead.
```suggestion
    mPollingThread = std::thread([&] { pollNOCCounters(); });
```

File: src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    while (handle != nullptr) {
+      // Determine the name of the device
+      struct xclDeviceInfo2 info;
+      xclGetDeviceInfo2(handle, &info);
+      std::string deviceName = std::string(info.mName);
+      mDevices.push_back(deviceName);
+
+      std::string outputFile = "noc_profile_" + deviceName + ".csv"; 
+      writers.push_back(new NOCProfilingWriter(outputFile.c_str(),
+			    deviceName.c_str(), index));
```
Comment (keryell) [inline src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp:49]: No TABs!

File: src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+   
+    // Just like HAL and power profiling, go through devices 
+    //  that exist and open a file for each
+    uint64_t index = 0;
+    void* handle = xclOpen(index, "/dev/null", XCL_INFO);
+    while (handle != nullptr) {
+      // Determine the name of the device
+      struct xclDeviceInfo2 info;
+      xclGetDeviceInfo2(handle, &info);
+      std::string deviceName = std::string(info.mName);
```
Comment (keryell) [inline src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp:44]:
```suggestion
      std::string deviceName { info.mName };
```

File: src/runtime_src/xdp/profile/plugin/aie/aie_plugin.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#ifndef XDP_AIE_PLUGIN_DOT_H
+#define XDP_AIE_PLUGIN_DOT_H
+
+#include <vector>
+#include <string>
+#include <thread>
```
Comment (keryell) [inline src/runtime_src/xdp/profile/plugin/aie/aie_plugin.h:22]: Sort

File: src/runtime_src/xdp/profile/plugin/aie/aie_plugin.cpp [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        // TEMPORARY: Larry is making AIE array available when xclbin gets loaded  
+        if (!aieArray) {
+          aieArray = new zynqaie::Aie(device);
+          drv->setAieArray(aieArray);
+        }
+
+        // Iterate over all AIE Counters
+        auto numCounters = (db->getStaticInfo()).getNumAIECounter(index);
+        for (uint64_t c=0; c < numCounters; c++) {
+          auto aie = (db->getStaticInfo()).getAIECounter(index, c);
```
Comment (keryell) [inline src/runtime_src/xdp/profile/plugin/aie/aie_plugin.cpp:115]: In this PR there *a lot* of useless `()` like here.

File: src/runtime_src/xdp/profile/device/aie_counter.h [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+ *
+ * This class only aims at providing interface for easy and
+ * safe access to a single profiling IP. Managing the
+ * association between IPs and devices should be done in a
+ * different data structure that is built on top of this class.
+ */
+class AIECounter : public ProfileIP {
+public:
+    /**
+     * The constructor takes a device handle and a ip index
```
Comment (keryell) [inline src/runtime_src/xdp/profile/device/aie_counter.h:44]:
```suggestion
     * The constructor takes a device handle and an ip index
```
also remove all the heading `*` in the comments

File: src/runtime_src/xdp/profile/database/dynamic_event_database.cpp [hunk 8]
```diff
@@ -193,4 +193,66 @@ namespace xdp {
 
     return powerSamples[deviceId] ;
   }
+
+  void VPDynamicDatabase::addAIESample(uint64_t deviceId, double timestamp,
+					 const std::vector<uint64_t>& values)
```
Comment (keryell) [inline src/runtime_src/xdp/profile/database/dynamic_event_database.cpp:198]: No TAbs!

**Review by keryell — CHANGES_REQUESTED** [review]

File: src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp (hunk 1 above)
Comment (pgschuey) [reply src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp]: fixed

File: src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp (hunk 2 above)
Comment (pgschuey) [reply src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp:62]: not addresses in this PR

File: src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp (hunk 3 above)
Comment (pgschuey) [reply src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp]: fixed

File: src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp (hunk 4 above)
Comment (pgschuey) [reply src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp:44]: not addressed

**Pull request merged**